Automattic/WP-Job-Manager
# Consume remote notices

Fixes #2374 

### Changes proposed in this Pull Request
* Started using `\WP_Job_Manager` namespace with new classes.
* Added `\WP_Job_Manager\WP_Job_Manager_Com_API` class to encapsulate all requests to wpjobmanager.com – currently only implementing notices consumption.
* Consuming notices endpoint and displaying those using the existing `render_notice` method.
* Added some `@deprecated` annotations to previous notices mechanism.
* Implemented `\WP_Job_Manager\Admin\Notices_Conditions` to encapsulate all conditions checking (added unit tests).

### Testing instructions
There is a couple of ways of testing this:
* Testing using your local wpjobmanager.com environment.
  * Check out your local wpjm.com environment to `feature/reengaging-customers`.
  * Apply the patch from [here](227-gh-Automattic/wpjobmanager.com) – you can edit the values to test the use case you want.
  * Make sure you are pointing to your local wpjobmanager.com environment.
  * Clean your local transients.
* Mocking the api responses with a snippet.
  * Use the following snippet:
```
add_filter( 'pre_http_request', function($preempt, $r, $url) {
	$parsed = parse_url( $url );

	if ( 'wpjobmanager.com' === $parsed['host'] ) {
		if ( '/wp-json/wpjmcom-notices/1.0/notices' === $parsed['path']) {
			return array(
            'headers'     => array(),
            'cookies'     => array(),
            'filename'    => null,
            'response'    => [ 'code' => 200 ],
            'status_code' => 200,
            'success'     => 1,
            'body'        => <<<EOQ
{
	"notices": {
		"test-notice-1": {
			"type": "site-wide",
			"icon": "wpjm",
			"heading": "Awesome Testing Header 1",
			"message": "This is a test notice message.",
			"actions": [
				{
					"label": "Read more",
					"url": "https:\/\/wpjobmanager.com",
					"primary": true,
					"target": "_blank"
				},
				{
					"label": "Other",
					"url": "https:\/\/wpjobmanager.com",
					"primary": false,
					"target": "_blank"
				}
			],
			"conditions": [
				{
					"type": "screens",
					"screens": [
						"wpjm*"
					]
				}
			]
		},
		"test-notice-2": {
			"type": "site-wide",
			"level": "warning",
			"icon": "wpjm",
			"heading": "Awesome Testing Header 2",
			"message": "This is a test warning message.",
			"actions": [
				{
					"label": "Read more",
					"url": "https:\/\/wpjobmanager.com",
					"primary": true,
					"target": "_blank"
				}
			],
			"conditions": [
				{
					"type": "screens",
					"screens": [
						"wpjm*"
					]
				}
			]
		},
		"test-notice-3": {
			"type": "user",
			"level": "error",
			"icon": "wpjm",
			"heading": "Awesome Testing Header 3",
			"message": "This is a test error message.",
			"actions": [
				{
					"label": "Read more",
					"url": "https:\/\/wpjobmanager.com",
					"primary": true,
					"target": "_blank"
				}
			],
			"conditions": [
				{
					"type": "screens",
					"screens": [
						"wpjm*"
					]
				}
			],
			"dismissible": false
		}

	}
}
EOQ,
        );
		}
	}
	
	return $preempt;
}, 9, 3 );
```
  * Clean your transients.
  * ⚠️ **If you are using a similar snippet to point to your local environment make sure that you have an early return there if `$preempt` is different from `false`**
* You might need to remove the `wp_job_manager_dismissed_notices` option to make the dismissed notices appear again.
* Test different options coming from the API, as well as conditions.


### Deprecated Code
* `WP_Job_Manager_Admin_Notices::add_notice` method.
* `WP_Job_Manager_Admin_Notices::remove_notice` method.
* `WP_Job_Manager_Admin_Notices::reset_notices` method.
* `WP_Job_Manager_Admin_Notices::has_notice` method.
* `job_manager_init_admin_notices` action hook.
* `job_manager_show_admin_notice_{notice_id}` filter.
* `job_manager_admin_notice_{notice_id}` action hook.

### New Hooks
* `job_manager_admin_notices` filter.

### Screenshot / Video
<img width="1653" alt="image" src="https://user-images.githubusercontent.com/799065/232599368-84ca867b-5c61-4de4-91ff-474ffec2d41b.png">
